### PR TITLE
Shortcodes: Merges the polldaddy shortcode changes from WP.com

### LINKED ISSUE #2079: HTTPS support in the Polldaddy shortcode
Howdy! Looks like http is hard-coded in various places in the Polldaddy shortcode, which prevents javascript from being loaded in many browsers on an HTTPS site.

### PR DESCRIPTION
**Pull request by zinigor**
This is a fairly large merge that brings the wpcom shortcode version in sync with the Jetpack version. Should fix #2079.

### CODE REVIEW

File: modules/shortcodes/polldaddy.php [hunk 1]
```diff
@@ -105,16 +185,20 @@ function polldaddy_shortcode( $atts ) {
...
+				$rating_js_file = "http://i0.poll.fm/js/rating/rating.js";
+
 			if ( $inline ) {
 				return <<<SCRIPT
 <div class="pd-rating" id="pd_rating_holder_{$rating}{$item_id}"></div>
 <script type="text/javascript" charset="UTF-8"><!--//--><![CDATA[//><!--
 PDRTJS_settings_{$rating}{$item_id}={$settings};
 //--><!]]></script>
-<script type="text/javascript" charset="UTF-8" src="http://i.polldaddy.com/ratings/rating.js"></script>
+<script type="text/javascript" charset="UTF-8" src="{$rating_js_file}"></script>
```
Comment (jeherve) [inline modules/shortcodes/polldaddy.php]: Shouldn't this be escaped with `esc_js()`?

File: modules/shortcodes/polldaddy.php [hunk 2]
```diff
@@ -214,21 +313,26 @@ function polldaddy_shortcode( $atts ) {
...
 			}
-		}
-		elseif ( !empty( $survey ) ) { //survey embed
+		} elseif ( !empty( $survey ) ) { //survey embed
 
 			if ( in_array( $type, array( 'iframe', 'button', 'banner', 'slider' ) ) ) {
 
 				if ( empty( $title ) ) {
-					$title = 'Take Our Survey';
+					$title = __( 'Take Our Survey' );
```
Comment (jeherve) [inline modules/shortcodes/polldaddy.php]: There is a missing Jetpack textdomain here.

File: modules/shortcodes/polldaddy.php [hunk 3]
```diff
@@ -297,31 +402,34 @@ function polldaddy_shortcode( $atts ) {
...
 	}
 
 	function generate_scripts() {
 		$script = '';
 
 		if ( is_array( self::$scripts ) ) {
+			if ( is_ssl() )
+				$rating_js_file = "https://polldaddy.com/js/rating/rating.js";
+			else
+				$rating_js_file = "http://i0.poll.fm/js/rating/rating.js";
```
Comment (jeherve) [inline modules/shortcodes/polldaddy.php]: Do we need to escape the URL used in js here as well?

File: modules/shortcodes/polldaddy.php [hunk 4]
```diff
@@ -13,12 +14,91 @@ class PolldaddyShortcode {
...
+	private function get_async_code( array $settings, $survey_link ) {
+		$embed_src     = 'http://i0.poll.fm/survey.js';
+		$embed_src_ssl = 'https://polldaddy.com/survey.js';
+
+		$include = <<<CONTAINER
+( function( d, c, j ) {
+  if ( !d.getElementById( j ) ) {
+    var pd = d.createElement( c ), s;
+    pd.id = j;
+    pd.src = ( 'https:' == d.location.protocol ) ? '{$embed_src_ssl}' : '{$embed_src}';
```
Comment (jeherve) [inline modules/shortcodes/polldaddy.php]: Should the URLs be escaped with `esc_js` here as well?

File: modules/shortcodes/polldaddy.php (hunk 1 above)
Comment (zinigor) [inline modules/shortcodes/polldaddy.php]: Couldn't hurt, but in this case we should be using `esc_url()`, shouldn't we? Anyway, thanks, will fix!

File: modules/shortcodes/polldaddy.php (hunk 2 above)
Comment (zinigor) [inline modules/shortcodes/polldaddy.php]: Good catch, thank you!

File: modules/shortcodes/polldaddy.php (hunk 4 above)
Comment (zinigor) [inline modules/shortcodes/polldaddy.php]: Yes, just the same, we may use `esc_url` just in case.

File: modules/shortcodes/polldaddy.php [hunk 5]
```diff
@@ -13,12 +14,91 @@ class PolldaddyShortcode {
...
+	
+			foreach ( $matches as $match ) {
+				if (!isset( $match[2] ) )
+					continue;
+					
+				$id = (int) $match[2];
+				
+				if ( $id > 0 ) {
+					$content = str_replace( $match[0], "[polldaddy poll=$id]", $content );
+					do_action( 'jetpack_embed_to_shortcode', 'polldaddy', $url );
```
Comment (kraftbj) [inline modules/shortcodes/polldaddy.php]: Need to add preceding
`/** This action is documented in modules/shortcodes/youtube.php */`

**Comment by zinigor** [comment]
@kraftbj Thanks for the doc hint, added.
@jeherve Thanks for spotting the missing textdomain, fixed. Also, I have been thinking about the escaping, and I don't see how it is useful here. We control the strings and we can see that they do not need any escaping. I wanted to add it just in case, but with HEREDOC notation it will be rather awkward, so let's just leave it like that.

File: modules/shortcodes/polldaddy.php [hunk 6]
```diff
@@ -13,12 +14,92 @@ class PolldaddyShortcode {
...
+    s = d.getElementsByTagName( c )[0];
+    s.parentNode.insertBefore( pd, s );
+  }
+}( document, 'script', 'pd-embed' ) );
+CONTAINER;
+
+		// Compress it a bit
+		$include = $this->compress_it( $include );
+
+		$placeholder = '<div class="pd-embed" data-settings="'.esc_attr( json_encode( $settings ) ).'"></div>';
```
Comment (dereksmart) [inline modules/shortcodes/polldaddy.php]: Should be whitespace around the string operators

File: modules/shortcodes/polldaddy.php [hunk 7]
```diff
@@ -13,12 +14,92 @@ class PolldaddyShortcode {
...
+		if ( $type === 'button' )
```
Comment (dereksmart) [inline modules/shortcodes/polldaddy.php]: Yoda conditionals plz

Comment (dereksmart) [inline modules/shortcodes/polldaddy.php]: Brackets for conditionals please

File: modules/shortcodes/polldaddy.php [hunk 8]
```diff
@@ -13,12 +14,92 @@ class PolldaddyShortcode {
...
+			$placeholder = '<a class="pd-embed" href="'.esc_attr( $survey_link ).'" data-settings="'.esc_attr( json_encode( $settings ) ).'">'.esc_html( $settings['title'] ).'</a>';
+
+		$js_include = $placeholder."\n";
+		$js_include .= '<script type="text/javascript"><!--//--><![CDATA[//><!--'."\n";
+		$js_include .= $include."\n";
```
Comment (dereksmart) [inline modules/shortcodes/polldaddy.php]: Should be whitespace around the string operators

File: modules/shortcodes/polldaddy.php [hunk 9]
```diff
@@ -13,12 +14,92 @@ class PolldaddyShortcode {
...
+CONTAINER;
+
+		// Compress it a bit
+		$include = $this->compress_it( $include );
+
+		$placeholder = '<div class="pd-embed" data-settings="'.esc_attr( json_encode( $settings ) ).'"></div>';
+		if ( $type === 'button' )
+			$placeholder = '<a class="pd-embed" href="'.esc_attr( $survey_link ).'" data-settings="'.esc_attr( json_encode( $settings ) ).'">'.esc_html( $settings['title'] ).'</a>';
+
+		$js_include = $placeholder."\n";
```
Comment (dereksmart) [inline modules/shortcodes/polldaddy.php]: Should be whitespace around the string operators

File: modules/shortcodes/polldaddy.php [hunk 10]
```diff
@@ -13,12 +14,92 @@ class PolldaddyShortcode {
...
+		$js_include .= '<script type="text/javascript"><!--//--><![CDATA[//><!--'."\n";
+		$js_include .= $include."\n";
+		$js_include .= "//--><!]]></script>\n";
+
+		if ( $type !== 'button' )
+			$js_include .= '<noscript>'.$survey_link."</noscript>\n";
```
Comment (dereksmart) [inline modules/shortcodes/polldaddy.php]: Should be whitespace around the string operators

File: modules/shortcodes/polldaddy.php [hunk 11]
```diff
@@ -13,12 +14,92 @@ class PolldaddyShortcode {
...
+		$placeholder = '<div class="pd-embed" data-settings="'.esc_attr( json_encode( $settings ) ).'"></div>';
+		if ( $type === 'button' )
+			$placeholder = '<a class="pd-embed" href="'.esc_attr( $survey_link ).'" data-settings="'.esc_attr( json_encode( $settings ) ).'">'.esc_html( $settings['title'] ).'</a>';
+
+		$js_include = $placeholder."\n";
+		$js_include .= '<script type="text/javascript"><!--//--><![CDATA[//><!--'."\n";
+		$js_include .= $include."\n";
+		$js_include .= "//--><!]]></script>\n";
+
+		if ( $type !== 'button' )
```
Comment (dereksmart) [inline modules/shortcodes/polldaddy.php]: Yoda conditionals & braces

File: modules/shortcodes/polldaddy.php [hunk 12]
```diff
@@ -13,12 +14,92 @@ class PolldaddyShortcode {
...
+	
+	/*
+	 * Polldaddy Poll Embed script - transforms code that looks like that:
+	 * <script type="text/javascript" charset="utf-8" src="http://static.polldaddy.com/p/123456.js"></script>
+	<noscript><a href="http://polldaddy.com/poll/123456/">What is your favourite color?</a></noscript>
+	 * into the [polldaddy poll=...] shortcode format
+	 */
+	function polldaddy_embed_to_shortcode( $content ) {
+				
+		if ( false === strpos( $content, 'polldaddy.com/p/' ) )
```
Comment (dereksmart) [inline modules/shortcodes/polldaddy.php]: missing braces

File: modules/shortcodes/polldaddy.php [hunk 13]
```diff
@@ -13,12 +14,92 @@ class PolldaddyShortcode {
...
+			return $content;
+
+		$regexes = array();
+
+		$regexes[] = '#<script[^>]+?src="https?://(secure|static)\.polldaddy\.com/p/([0-9]+)\.js"[^>]*+>\s*?</script>\r?\n?(<noscript>.*?</noscript>)?#i';
+
+		$regexes[] = '#&lt;script(?:[^&]|&(?!gt;))+?src="https?://(secure|static)\.polldaddy\.com/p/([0-9]+)\.js"(?:[^&]|&(?!gt;))*+&gt;\s*?&lt;/script&gt;\r?\n?(&lt;noscript&gt;.*?&lt;/noscript&gt;)?#i';
+	
+		foreach ( $regexes as $regex ) {
+			if ( ! preg_match_all( $regex, $content, $matches, PREG_SET_ORDER ) )
```
Comment (dereksmart) [inline modules/shortcodes/polldaddy.php]: Braces

File: modules/shortcodes/polldaddy.php [hunk 14]
```diff
@@ -13,12 +14,92 @@ class PolldaddyShortcode {
...
+		 		continue;
+	
+			foreach ( $matches as $match ) {
+				if (!isset( $match[2] ) )
```
Comment (dereksmart) [inline modules/shortcodes/polldaddy.php]: missing whitespace

File: modules/shortcodes/polldaddy.php [hunk 15]
```diff
@@ -105,16 +186,20 @@ function polldaddy_shortcode( $atts ) {
 
 			$item_id = esc_js( $item_id );
 
+			if ( is_ssl() )
```
Comment (dereksmart) [inline modules/shortcodes/polldaddy.php]: Braces please

File: modules/shortcodes/polldaddy.php [hunk 16]
```diff
@@ -181,26 +258,49 @@ function polldaddy_shortcode( $atts ) {
 							$margins = 'margin: 0px 0px 0px 10px';
 					}
 
+					// Force the normal style embed on single posts/pages otherwise it's not rendered on infinite scroll themed blogs ('infinite_scroll_render' isn't fired)
+					if ( is_singular() )
```
Comment (dereksmart) [inline modules/shortcodes/polldaddy.php]: Braces

File: modules/shortcodes/polldaddy.php [hunk 17]
```diff
@@ -181,26 +258,49 @@ function polldaddy_shortcode( $atts ) {
...
+						$inline = true;
+
 					if ( $cb === false && !$inline ) {
 						if ( self::$scripts === false )
 							self::$scripts = array();
 
 						$data = array( 'url' => $poll_js );
 
-						self::$scripts['poll'][] = $data;
+						self::$scripts['poll'][intval( $poll )] = $data;
```
Comment (dereksmart) [inline modules/shortcodes/polldaddy.php]: missing spaces in the $poll bracket.  Also should we first define this `$poll = intval( $poll )`?

File: modules/shortcodes/polldaddy.php [hunk 18]
```diff
@@ -297,31 +403,34 @@ function polldaddy_shortcode( $atts ) {
...
+		} else {
 			return '<!-- no polldaddy output -->';
+		}
 	}
 
 	function generate_scripts() {
 		$script = '';
 
 		if ( is_array( self::$scripts ) ) {
+			if ( is_ssl() )
```
Comment (dereksmart) [inline modules/shortcodes/polldaddy.php]: Braces

File: modules/shortcodes/polldaddy.php (hunk 3 above)
Comment (dereksmart) [inline modules/shortcodes/polldaddy.php]: I don't believe that is necessary here

File: modules/shortcodes/polldaddy.php [hunk 19]
```diff
@@ -297,31 +403,34 @@ function polldaddy_shortcode( $atts ) {
...
 				}
-				return <<<CONTAINER
-<script type="text/javascript" charset="UTF-8" src="http://i0.poll.fm/survey.js"></script>
-<script type="text/javascript" charset="UTF-8"><!--//--><![CDATA[//><!--
-polldaddy.add( {$settings} );
-//--><!]]></script>
-<noscript>{$survey_link}</noscript>
-CONTAINER;
+
+				if ( empty( $settings ) )
```
Comment (dereksmart) [inline modules/shortcodes/polldaddy.php]: Braces

File: modules/shortcodes/polldaddy.php [hunk 20]
```diff
@@ -214,21 +314,26 @@ function polldaddy_shortcode( $atts ) {
...
 			if ( in_array( $type, array( 'iframe', 'button', 'banner', 'slider' ) ) ) {
 
 				if ( empty( $title ) ) {
-					$title = 'Take Our Survey';
+					$title = __( 'Take Our Survey' );
 					if( !empty( $link_text ) )
 						$title = $link_text;
 				}
 
+				if ( $type == 'banner' || $type == 'slider' )
```
Comment (dereksmart) [inline modules/shortcodes/polldaddy.php]: Braces

**Comment by dereksmart** [comment]
I can say I have tested it, and the shortcode works properly.  The code has a whole lot of code style problems, and lacks documentation.  That said I don't see any blatant security holes or anything.  I agree with @zinigor about escaping those URLs.  Those are only set by us in the code, so no real need there.

**Comment by zinigor** [comment]
@dereksmart thanks for the review! I have fixed the issues you pointed out, the file should be fully style compliant now.

**Comment by dereksmart** [comment]
Looks great! tested with some of my own polldaddy shortcodes and works well.

File: modules/shortcodes/polldaddy.php [hunk 21]
```diff
@@ -13,12 +14,107 @@ class PolldaddyShortcode {
...
+				if ( ! isset( $match[2] ) ) {
+					continue;
+				}
+
+				$id = (int) $match[2];
+
+				if ( $id > 0 ) {
+					$content = str_replace( $match[0], "[polldaddy poll=$id]", $content );
+					/** This action is documented in modules/shortcodes/youtube.php */
+					do_action( 'jetpack_embed_to_shortcode', 'polldaddy', $url );
```
Comment (dereksmart) [inline modules/shortcodes/polldaddy.php]: $url throws an `undefined` notice when using the embed script

**Comment by zinigor** [comment]
@dereksmart Thanks for taking a look, I have fixed the bug you mentioned, should be good to go now.

**Pull request merged**